### PR TITLE
Add docs for ISymbol and its derived interfaces that say that they shoul...

### LINKED ISSUE #632: Docs for ISymbol and its derived interfaces should say that they are not to be extended outside Roslyn
The interface ISymbol and its derived interfaces should contain documentation that they are not intended to be extensible in client code. We intend to evolve these interfaces by the addition of members in future revisions (e.g. the addition of ILocalSymbol.IsReadonly), and we need to avoid having clients inherit or implement the interfaces to avoid breaking compatibility.

I recommend something like the following paragraph.

```cs
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
```

### PR DESCRIPTION
**Pull request by pdelvo**
...d not be implemented outside Roslyn.

Fixes #632.

### CODE REVIEW

**Comment by mattwar** [comment]
Is having this in a remarks section good enough? I don't think anyone will ever see these unless they are reading the source.

**Comment by pdelvo** [comment]
I think the question should ne: Why is Visual Studios rendering of xml docs that minimal? You can really put a lot of work into them just to be disappointed about the number of tags VS actually supports. Why doesn't the remarks section show up? Why isn't <list>, <code> supported? I think that would really improve productivity if the documentation tool tip would support that. I will move that sentence into the summary element when I get home but that really feels like an workaround.

**Comment by pdelvo** [comment]
I moved the text into the summary. I might open an issue for better support for advanced xml docs.

**Comment by pharring** [comment]
@dotnet-bot retest this, please.

**Comment by pharring** [comment]
@gafter I seem to remember you had an idea for how to address the general problem of tagging/enforcing public interfaces 'for internal implementation only'?

**Comment by gafter** [comment]
@pharring Yes, and it was not approved either as a language change or as a compiler change. There will be an analyzer that helps enforce it.

**Comment by pharring** [comment]
Ah, I see it right there on ISymbol: InternalImplementationOnlyAttribute. And it's enforced via an analyzer.
Is the attribute on the base interface sufficient for the analyzer to catch all cases?

**Comment by gafter** [comment]
@mattwar Do you prefer this to the version where it was in the remarks section.

We will also have an analyzer to "enforce" this.

**Comment by pharring** [comment]
Once again answering my own question: yes. The unit tests for the analyzer check the derived cases.
If this passes verification, then it's safe to merge.

**Comment by gafter** [comment]
@pharring I believe we are going to get rid of the attribute and just enforce it for these types directly.

**Comment by sharwell** [comment]
:exclamation: I do not think we should be adding a large amount of text to the summary line for commonly used types for a condition that developers would rarely (at best) consider violating. If this was SHFB-style documentation, it would be inside the `<remarks>` section, and wrapped:

```
/// <remarks>
/// <note type="implement">
/// <para>{warning here}</para>
/// </note>
/// </remarks>
```

Regardless of the exact form, I believe this should be in the `<remarks>` section, and placed below any remarks related to the valid use of the type.

**Comment by sharwell** [comment]
It makes me sad to think how little of the information I wrote in the documentation for [`CoreTaskExtensions`](https://github.com/rackerlabs/dotnet-threading/blob/master/Rackspace.Threading/CoreTaskExtensions.cs) is ever exposed to users. :cry:

**Comment by gafter** [comment]
@pdelvo Can you please put this back in the <remarks> section (as described in the corresponding issue)?

**Comment by pdelvo** [comment]
I pushed my old version again with the docs being in the remarks section.

File: src/Compilers/Core/Portable/Symbols/ISynthesizedMethodBodyImplementationSymbol.cs [hunk 1]
```diff
@@ -5,6 +5,10 @@ namespace Microsoft.CodeAnalysis
     /// <summary>
     /// Synthesized symbol that implements a method body feature (iterator, async, lambda, etc.)
     /// </summary>
+    /// <remarks>
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Symbols/ISynthesizedMethodBodyImplementationSymbol.cs]: :question: This interface is `internal`. Does it still need this note?

**Comment by sharwell** [comment]
It's too bad we can't use the `inheritdoc` or `token` constructs from Sandcastle Help File Builder to simplify this. :disappointed:

File: src/Compilers/Core/Portable/Symbols/ISynthesizedMethodBodyImplementationSymbol.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Symbols/ISynthesizedMethodBodyImplementationSymbol.cs]: Don't care.

**Pull request merged**